### PR TITLE
Fix pre-commit and linter versions to avoid surprise

### PR DESCRIPTION
**Pull request by li-boxuan**
We have a lot of headache with linting this project. Although https://github.com/OpenDevin/OpenDevin/pull/1071 made a first attempt to unify the behavior of local linting and CI linting, it still has the following limitations:

1) The version of pre-commit run locally (pyproject.toml) and that in CI config (lint.yml) are not fixed, and can diverge.
2) The versions of linters (ruff and mypy) are not fixed, which means new releases of these libraries can bring us surprises.

To avoid surprising results on GitHub Actions, e.g. a new release of pre-commit starts to reject all PRs, fix it to the latest version, 3.7.0, in both pyproject.toml and lint.yml. This PR also fixes ruff and mypy versions in pyproject.toml since we very likely don't really need latest upgrades from linters.

### CODE REVIEW

**Review by neubig — APPROVED** [review]
Thank you!

**Comment by li-boxuan** [comment]
@neubig Thanks for the quick review. Looks like I need you to sign-off once more since I pushed an one-liner after your approval.

**Review by enyst — APPROVED** [review]

**Pull request merged**
